client/session.go

```diff
@@ -3131,23 +3130,32 @@ func (b *baseBlocksResult) newDatabaseBlock(block *rpc.Block) (block.DatabaseBlo
 	switch {
 	case segments.Merged != nil:
 		// Unmerged, can insert directly into a single block
-		result.Reset(start, b.segmentForBlock(segments.Merged))
+		result.Reset(start, blockSize, b.segmentForBlock(segments.Merged))
 
 	case segments.Unmerged != nil:
 		// Must merge to provide a single block
 		segmentReaderPool := b.blockOpts.SegmentReaderPool()
-		readers := make([]io.Reader, len(segments.Unmerged))
-		for i := range segments.Unmerged {
+		readers := make([]xio.SegmentReader, len(segments.Unmerged))
+		end := time.Time{}
+		for i, segment := range segments.Unmerged {
 			segmentReader := segmentReaderPool.Get()
-			segmentReader.Reset(b.segmentForBlock(segments.Unmerged[i]))
-			readers[i] = segmentReader
+			endTime := timeConvert(segment.EndTime)
+			blockReader := xio.BlockReader{
+				SegmentReader: segmentReader,
+				Start:         timeConvert(segment.StartTime),
+				End:           endTime,
+			}
+			blockReader.Reset(b.segmentForBlock(segment))
+			readers[i] = blockReader
```

Comment:
Hm did this code need to change?  It seems that we'll need to heap allocate the `xio.BlockReader` here.  You're putting each element into `readers` which is a `[]xio.SegmentReader` yeah?  Maybe we can just directly put the segment readers in that array rather than wrapping them in a `BlockReader` unnecessarily?  They're going to get merged right below on line 3154 anyhow and will be discarded.